extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/oauth/OAuth2IdentityManager.java

```diff
@@ -0,0 +1,77 @@
+package io.quarkus.elytron.security.oauth2.runtime.oauth;
+
+import java.util.List;
+
+import org.jboss.logging.Logger;
+import org.wildfly.security.auth.server.RealmUnavailableException;
+import org.wildfly.security.auth.server.SecurityDomain;
+import org.wildfly.security.auth.server.SecurityIdentity;
+import org.wildfly.security.evidence.BearerTokenEvidence;
+
+import io.undertow.security.idm.Account;
+import io.undertow.security.idm.Credential;
+import io.undertow.security.idm.IdentityManager;
+
+public class OAuth2IdentityManager implements IdentityManager {
+    private static Logger log = Logger.getLogger(OAuth2IdentityManager.class);
+    private final SecurityDomain domain;
+    private String roleClaim;
+    private String roleSeparator;
+
+    public OAuth2IdentityManager(SecurityDomain domain, String roleClaim, String roleSeparator) {
+        this.domain = domain;
+        this.roleClaim = roleClaim;
+        // configs are trimmed so we use ' ' instead of a space string as default value
+        this.roleSeparator = roleSeparator == "' '" ? " " : roleSeparator;
+    }
+
+    @Override
+    public Account verify(Account account) {
+        return account;
+    }
+
+    @Override
+    public Account verify(String id, Credential credential) {
+        return verify(null, credential);
+    }
+
+    @Override
+    public Account verify(Credential credential) {
+        log.debugf("verify, credential=%s", credential);
+        try {
+            if (credential instanceof Oauth2Credential) {
+                Oauth2Credential oauth2Credential = (Oauth2Credential) credential;
+
+                try {
+                    BearerTokenEvidence evidence = new BearerTokenEvidence(oauth2Credential.getBearerToken());
+                    SecurityIdentity result = domain.authenticate(evidence);
+                    String[] roles = extractRoles(result);
+
+                    log.debugf("authenticate, result=%s", result);
+                    if (result != null) {
+                        return new OAuth2Account(result, roles);
+                    }
+                } catch (RealmUnavailableException e) {
+                    log.debugf(e, "failed, credential=%s", credential);
+                }
+            }
+        } catch (Exception e) {
+            log.warnf(e, "Failed to verify credential=%s", credential);
+        }
+        return null;
+    }
+
+    private String[] extractRoles(SecurityIdentity result) {
+        ElytronOAuth2CallerPrincipal principal = (ElytronOAuth2CallerPrincipal) result.getPrincipal();
+        if (roleSeparator == null) {
+            //we assume the claim is an array, so it will be mapped to a list
+            return ((List<String>) principal.getClaims()).toArray(new String[0]);
```

Comment:
@loicmathieu see the comment about `rolesSeparator`. And I'm actually not sure this code works. If the user does not set `rolesSeparator` then it just casts all the claims to List but they are returned as a Map from the Principal.
And having to set `rolesSeparator` for the simple role claim to work does seems like an extra requirement.
I propose to check if the roleClaim is a List then just use it as is, else if it is a String - wrap it in a List, unless it is `scope` which is a specific case

Reply:
Yes, it was not workin :)
I remove the rolesSeparator and implements a check if it's a list: return it, else split by space and return the list.

Reply:
yeah, that just would work for now :-), not even necessary to check if it 'scope' - will do the same in smallrye-jwt